lib/classes/Variables.js

```diff
@@ -0,0 +1,292 @@
+'use strict';
+
+const _ = require('lodash');
+const path = require('path');
+const traverse = require('traverse');
+const replaceall = require('replaceall');
+
+class Variables {
+
+  constructor(serverless) {
+    this.serverless = serverless;
+    this.service = this.serverless.service;
+    this.variableSyntax = RegExp(this.service.defaults.variableSyntax, 'g');
+    this.fileRefSyntax = RegExp(/^file\(([a-zA-Z0-9._\-\/]+?)\)/g);
+  }
+
+  populateService(processedOptions) {
+    const that = this;
+    this.options = processedOptions || {};
+    const variableSyntaxProperty = this.service.defaults.variableSyntax;
+
+    // temporally remove variable syntax from service otherwise it'll match
+    this.service.defaults.variableSyntax = true;
+
+    /*
+     * we can't use an arrow function in this case cause that would
+     * change the lexical scoping required by the traverse module
+     */
+    traverse(this.service).forEach(function (propertyParam) {
+      const t = this;
+      let property = propertyParam;
+
+      if (typeof property === 'string') {
+        property = that.populateProperty(property);
+        t.update(property);
+      }
+    });
+
+    this.service.defaults.variableSyntax = variableSyntaxProperty;
+    console.log(this.service.custom)
+    return this.service;
+  }
+
+  populateProperty(propertyParam) {
+    let property = propertyParam;
+    if (typeof property === 'string' && property.match(this.variableSyntax)) {
+      property.match(this.variableSyntax).forEach((matchedString) => {
+        let variableString = matchedString
+          .replace(this.variableSyntax, (match, varName) => varName.trim());
+
+        // remove spaces
+        variableString = variableString.replace(/\s/g, '');
+
+        if (variableString.match(RegExp(/,/g))) {
+          property = this.overwrite(variableString, property, matchedString);
+        } else if (variableString.match(RegExp(/^opt\..*/g))) {
```

Comment:
matching for the different variable types is duplicated several times here. Needs to be in a separate method. Imho best case in populateProperty you only check if it has an override or not, if not call a method that will figure out what type of variable it is and then call the corresponding method. If it is override it calls an override method, the override method calls the before mentioned variable method and if it returns a non empty string it returns it, otherwise it calls the same function with the second variable definition